pkg/schema/definitions/converter.go

```diff
@@ -0,0 +1,223 @@
+package definitions
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/rancher/apiserver/pkg/types"
+	wapiextv1 "github.com/rancher/wrangler/v2/pkg/generated/controllers/apiextensions.k8s.io/v1"
+	wranglerDefinition "github.com/rancher/wrangler/v2/pkg/schemas/definition"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+var (
+	ErrNotFound     = errors.New("not found")
+	ErrNotRefreshed = errors.New("not refreshed")
+)
+
+// crdToDefinition builds a schemaDefinition for a CustomResourceDefinition
+func crdToDefinition(crdCache wapiextv1.CustomResourceDefinitionCache, crdName string, modelName string, version string) (schemaDefinition, error) {
+	crd, err := crdCache.Get(crdName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return schemaDefinition{}, ErrNotFound
+		}
+		return schemaDefinition{}, err
+	}
+
+	var jsonSchemaProps *apiextv1.JSONSchemaProps
+	for _, crdVersion := range crd.Spec.Versions {
+		if crdVersion.Name == version {
+			jsonSchemaProps = crdVersion.Schema.OpenAPIV3Schema
+			break
+		}
+	}
+
+	if jsonSchemaProps == nil {
+		return schemaDefinition{}, fmt.Errorf("unknown version %q for CRD %q", version, crdName)
+	}
+
+	// CRD definitions generally has more information than the OpenAPI V2
+	// because it embeds an OpenAPI V3 document. However, these 3 fields
+	// are the exception where the Open API V2 endpoint has more
+	// information.
+	//
+	// To avoid overriding these later on, we remove them here. Yeah, really.
+	delete(jsonSchemaProps.Properties, "apiVersion")
+	delete(jsonSchemaProps.Properties, "kind")
+	delete(jsonSchemaProps.Properties, "metadata")
+
+	path := proto.NewPath(modelName)
+
+	definitions := make(map[string]definition)
+	convertJSONSchemaPropsToDefinition(*jsonSchemaProps, path, definitions)
+
+	return schemaDefinition{
+		DefinitionType: modelName,
+		Definitions:    definitions,
+	}, nil
+}
+
+func convertJSONSchemaPropsToDefinition(props apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) {
```

Comment:
> Personally speaking, I'd prefer to convert everything to the proto.Schema type and then allow the schema to process things. 

Agreed. Though with the bug found in the proto.Schema handling for OpenAPI V3, I think we want to avoid the following path: CRD -> JSONSchema (json raw bytes) -> proto.Schema. (eg: Fields with references won't work correctly).

Another approach could be for us to do the conversion from CRD to proto schema directly. I can give that a go.

> As a nit/aside, I think that this function is complex enough to use it's own comment.

Will do.